### PR TITLE
perf: speed up when startup meet timeout

### LINKED ISSUE #60: feature request: speed up user's startup time as fast as possible
**Is your feature request related to a problem? Please describe.**
When first time load a namespace from remote, 

we will get
* 404: If namespace doesn't exists in remote, apollo-client will get 404,  when first time load this namespace will be slow. fix in https://github.com/apolloconfig/apollo-java/pull/61 already, will release with 2.3.0 in the feture.
* timeout: if apollo-configservice down, the loading of namespace will slow too.

load 10 namespaces, timeusage
* **11494 ms** when all of them meet **404**
* **65929 ms** when all of them meet **timeout**

**Describe the solution you'd like**

* decrease the total count of io 
* fast fail when meet error

**Additional context**

### 404

The code to simulate 404 when load namespace.

Main.java
```java
import com.ctrip.framework.apollo.Config;
import com.ctrip.framework.apollo.ConfigService;
import java.util.ArrayList;
import java.util.List;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class Main {

  private static final Logger log = LoggerFactory.getLogger(Main.class);

  public static void main(String[] args) {
    System.setProperty("app.id", "custom");
    System.setProperty("apollo.meta", "http://81.68.181.139:8080");
    // System.setProperty("apollo.loadConfigQPS", "5");
    runAndCountTime("load multiple namespace", () -> load());
  }

  static void load() {
    List<String> list = new ArrayList<>();
    for (int i = 1; i <= 10; i++) {
      list.add("ns" + i);
    }
    log.info("try to load {} namespace in list {}", list.size(), list);

    for (String ns : list) {
      runAndCountTime("load namespace " + ns, () -> {
        Config config = ConfigService.getConfig(ns);
        log.info("load namespace '{}' finished. k1 = {}", ns, config.getProperty("k1", null));
      });
    }
  }

  static void runAndCountTime(String message, Runnable runnable) {
    log.info("start {}", message);
    long s = System.currentTimeMillis();
    runnable.run();
    long end = System.currentTimeMillis();
    log.info("end {} time use {} ms", message, end - s);
  }
}
```

the output log 
[before-change.log](https://github.com/apolloconfig/apollo-java/files/15237177/before-change.log)

### timeout

The code to simulate timeout when load namespace.

Timeout.java
```java
import com.ctrip.framework.apollo.Config;
import com.ctrip.framework.apollo.ConfigService;
import java.util.ArrayList;
import java.util.List;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class Timeout {

  private static final Logger log = LoggerFactory.getLogger(Timeout.class);

  public static void main(String[] args) {
    System.setProperty("app.id", "custom");
    // System.setProperty("apollo.meta", "http://81.68.181.139:8080");
    // wrong meta address will cause timeout
    System.setProperty("apollo.meta", "http://81.68.181.100:8080");
    // System.setProperty("apollo.loadConfigQPS", "5");
    runAndCountTime("load multiple namespace", () -> load());
  }

  static void load() {
    List<String> list = new ArrayList<>();
    for (int i = 1; i <= 10; i++) {
      list.add("ns" + i);
    }
    log.info("try to load {} namespace in list {}", list.size(), list);

    for (String ns : list) {
      runAndCountTime("load namespace " + ns, () -> {
        Config config = ConfigService.getConfig(ns);
        log.info("load namespace '{}' finished. k1 = {}", ns, config.getProperty("k1", null));
      });
    }
  }

  static void runAndCountTime(String message, Runnable runnable) {
    log.info("start {}", message);
    long s = System.currentTimeMillis();
    runnable.run();
    long end = System.currentTimeMillis();
    log.info("end {} time use {} ms", message, end - s);
  }
}
```

the output log 
[timeout-before-change.log](https://github.com/apolloconfig/apollo-java/files/15237371/timeout-before-change.log)

### PR DESCRIPTION
**Pull request by Anilople**
## What's the purpose of this PR

Speed up when startup meet timeout, i.e when meet meta service and config service crash.

## Which issue(s) this PR fixes:
Fixes #60 

## Brief changelog

There is a synchronized lock in 

https://github.com/apolloconfig/apollo-java/blob/3f0979da3ea6478304719a71d4991fe4787558d1/apollo-client/src/main/java/com/ctrip/framework/apollo/internals/ConfigServiceLocator.java#L193

It will block even use multiple thread.

When try to getConfigServices, if cache is empty, method `updateConfigServices` will be invoke

https://github.com/apolloconfig/apollo-java/blob/3f0979da3ea6478304719a71d4991fe4787558d1/apollo-client/src/main/java/com/ctrip/framework/apollo/internals/ConfigServiceLocator.java#L162-L167

In this pr, the code change to **fail quickly**, so it can speed up when meta service and config service crash(that will cause timeout for client).

Maybe it better to let daemon thread to invoke method `updateConfigServices`, 
another thread just use the result in memory cache,
If there is no available config service, throw exception quickly.

After change, load 10 namespaces timeusage 
from **65929 ms to 5150ms** when all of them meet timeout

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Read the [Contributing Guide](https://github.com/apolloconfig/apollo/blob/master/CONTRIBUTING.md) before making this pull request.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit tests to verify the code.
- [x] Run `mvn clean test` to make sure this pull request doesn't break anything.
- [x] Update the [`CHANGES` log](https://github.com/apolloconfig/apollo-java/blob/master/CHANGES.md).

## Summary by CodeRabbit

- **New Features**
  - Enhanced startup performance to handle timeout situations more efficiently in Apollo Java 2.3.0.

- **Bug Fixes**
  - Improved handling of configuration services with rate limiting and timeout settings.

- **Tests**
  - Added new test methods and assertions to ensure robust error handling and configuration updates.

- **Documentation**
  - Updated CHANGES.md to reflect performance enhancements in the Apollo Java 2.3.0 release.

### CODE REVIEW

File: apollo-client/src/main/java/com/ctrip/framework/apollo/internals/ConfigServiceLocator.java [hunk 1]
```diff
@@ -160,8 +168,14 @@ private String getDeprecatedCustomizedConfigService() {
...
   public List<ServiceDTO> getConfigServices() {
+    initialize();
     if (m_configServices.get().isEmpty()) {
-      updateConfigServices();
+      // quick fail
+      throw new ApolloConfigException(
+          "No available config service, "
+              + "server side maybe crash or network cannot connect to server from this ip, "
+          + "one of meta service url is " + assembleMetaServiceUrl()
+      );
```
Comment (nobodyiam) [inline apollo-client/src/main/java/com/ctrip/framework/apollo/internals/ConfigServiceLocator.java:178]: Is it the same if we keep all the logic unchanged and only remove the `updateConfigServices` call and throw the exception here?

File: apollo-client/src/main/java/com/ctrip/framework/apollo/internals/ConfigServiceLocator.java [hunk 2]
```diff
@@ -194,6 +200,11 @@ private synchronized void updateConfigServices() {
     String url = assembleMetaServiceUrl();
 
     HttpRequest request = new HttpRequest(url);
+    // speed up http request
+    int mixTimeout = Math.min(m_configUtil.getConnectTimeout(), m_configUtil.getReadTimeout());
+    request.setConnectTimeout(mixTimeout);
+    request.setReadTimeout(mixTimeout);
```
Comment (Anilople) [inline apollo-client/src/main/java/com/ctrip/framework/apollo/internals/ConfigServiceLocator.java:206]: Use min timeout here.

File: apollo-client/src/test/java/com/ctrip/framework/apollo/integration/ConfigIntegrationTest.java [hunk 3]
```diff
@@ -60,51 +46,9 @@
...
 public class ConfigIntegrationTest extends BaseIntegrationTest {
 
-  private String someReleaseKey;
-  private File configDir;
-  private String defaultNamespace;
-  private String someOtherNamespace;
-  private RemoteConfigLongPollService remoteConfigLongPollService;
-
-  @BeforeEach
-  public void setUp() throws Exception {
```
Comment (Anilople) [inline apollo-client/src/test/java/com/ctrip/framework/apollo/integration/ConfigIntegrationTest.java:70]: move those code to parent class

File: apollo-client/src/test/java/com/ctrip/framework/apollo/BaseIntegrationTest.java [hunk 4]
```diff
@@ -108,10 +120,22 @@ public void setUp() throws Exception {
...
+      configDir.delete();
+    }
+    configDir.mkdirs();
   }
 
   @AfterEach
   public void tearDown() throws Exception {
+    // get the instance will trigger long poll task execute, so move it from setup to tearDown
+    RemoteConfigLongPollService remoteConfigLongPollService
+        = ApolloInjector.getInstance(RemoteConfigLongPollService.class);
```
Comment (Anilople) [inline apollo-client/src/test/java/com/ctrip/framework/apollo/BaseIntegrationTest.java:135]: `ApolloInjector.getInstance(RemoteConfigLongPollService.class);` will trigger context load, so move it to `tearDown` method

File: apollo-client/src/main/java/com/ctrip/framework/apollo/internals/ConfigServiceLocator.java (hunk 1 above)
Comment (Anilople) [reply apollo-client/src/main/java/com/ctrip/framework/apollo/internals/ConfigServiceLocator.java]: Done.

File: apollo-client/src/main/java/com/ctrip/framework/apollo/internals/ConfigServiceLocator.java [hunk 5]
```diff
@@ -161,7 +161,13 @@ private String getDeprecatedCustomizedConfigService() {
    */
   public List<ServiceDTO> getConfigServices() {
     if (m_configServices.get().isEmpty()) {
-      updateConfigServices();
+      m_executorService.submit(() -> this.tryUpdateConfigServices());
```
Comment (Anilople) [inline apollo-client/src/main/java/com/ctrip/framework/apollo/internals/ConfigServiceLocator.java:164]: The default `scheduleAtFixedRate` is too late(5 minutes), 
I think we need to speed up it by submit immediately.

But that will make multiple http request.

There are some solutions:
* Solution 1: Add a rate limiter here
* Solution 2: scheduleAtFixedRate another method `recover` 5 seconds period, use a boolean to mark execute or not, and set the boolean to true in this line.
* Solution 3: Use a state machine to indicate task execute or not, if in executing, then will not submit task.
 
Or use another solution?

File: apollo-client/src/main/java/com/ctrip/framework/apollo/internals/ConfigServiceLocator.java (hunk 2 above)
Comment (nobodyiam) [reply apollo-client/src/main/java/com/ctrip/framework/apollo/internals/ConfigServiceLocator.java:206]: The connection time and read timeout serve different functions, so using the minimum of the two may not be ideal.

File: apollo-client/src/main/java/com/ctrip/framework/apollo/internals/ConfigServiceLocator.java (hunk 5 above)
Comment (nobodyiam) [reply apollo-client/src/main/java/com/ctrip/framework/apollo/internals/ConfigServiceLocator.java:164]: A rate limiter should be effective.

Comment (Anilople) [reply apollo-client/src/main/java/com/ctrip/framework/apollo/internals/ConfigServiceLocator.java]: Done, use default 1 qps.

add 3 config

```
apollo.discoveryQPS
apollo.discoveryConnectTimeout
apollo.discoveryReadTimeout
```

let user can custom behavior of discovery

**Review by nobodyiam — APPROVED** [review]
LGTM

**Pull request merged**